lib/Analysis/FIRRTLInstanceInfo.cpp

```diff
@@ -0,0 +1,156 @@
+
+void InstanceInfo::LatticeValue::merge(LatticeValue that) {
+  if (kind > that.kind)
+    return;
+
+  if (kind < that.kind) {
+    kind = that.kind;
+    constant = that.constant;
+    return;
+  }
+
+  if (kind == Constant && constant != that.constant)
+    kind = Mixed;
+}
+
+void InstanceInfo::LatticeValue::merge(bool property) {
+  merge({/*kind=*/Constant, /*constant=*/property});
+}
+
+namespace {
+/// This struct is one frame of a worklist.  This is used to track what instance
+/// is being visited and with what information about its instantiation needs to
+/// be known.
+struct Frame {
+
+  /// The instance that will be visited.
+  InstanceGraphNode *node;
+
+  /// Indicates that the current hierarchy is under the DUT.
+  bool underDut = false;
+
+  /// Indicates that the current hierarchy is under a layer.
+  bool underLayer = false;
+};
+} // namespace
+
+InstanceInfo::InstanceInfo(Operation *op, mlir::AnalysisManager &am) {
+  auto &iGraph = am.getAnalysis<InstanceGraph>();
+
+  SmallVector<Frame> worklist({{iGraph.getTopLevelNode()}});
+  while (!worklist.empty()) {
+    auto frame = worklist.pop_back_val();
+    auto *node = frame.node;
+    auto moduleOp = node->getModule();
+
+    // Compute information about this instance.
+    bool isDut = AnnotationSet(moduleOp).hasAnnotation(dutAnnoClass);
+
+    // Set the baseline attributes.  This is what the attributes will be if
+    // there is no other, existing attributes.
+    ModuleAttributes attributes({/*isDut=*/isDut});
+    attributes.underDut.merge(frame.underDut);
+    attributes.underLayer.merge(frame.underLayer);
+
+    // Merge the baseline attributes with the existing attributes, if they
+    // exist.
+    auto it = moduleAttributes.find(moduleOp);
+    if (it == moduleAttributes.end())
+      moduleAttributes[moduleOp] = attributes;
+    else {
+      auto &oldAttributes = it->getSecond();
+      oldAttributes.underDut.merge(attributes.underDut);
+      oldAttributes.underLayer.merge(attributes.underLayer);
+      attributes = oldAttributes;
+    }
+
+    for (auto *inst : *frame.node) {
+      auto underDut = frame.underDut || isDut;
+      auto underLayer = frame.underLayer ||
+                        inst->getInstance()->getParentOfType<LayerBlockOp>();
+      worklist.push_back({inst->getTarget(),
+                          /*underDut=*/underDut,
+                          /*underLayer=*/underLayer});
+    }
+  }
+  for (auto *node : llvm::depth_first(iGraph.getTopLevelNode())) {
+    auto moduleOp = node->getModule();
+
+    auto &attributes = moduleAttributes[moduleOp];
+
+    if (AnnotationSet(moduleOp).hasAnnotation(dutAnnoClass)) {
+      attributes.isDut = true;
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "InstanceInfo Analysis Results:\n";
+    DenseSet<Operation *> visited;
+    for (auto *node : llvm::depth_first(iGraph.getTopLevelNode())) {
+      auto moduleOp = node->getModule();
+      if (!visited.insert(moduleOp.getOperation()).second)
```

Comment:
Checking if already visited shouldn't be necessary, although using the variant of the iterator that stores the set externally is useful sometimes but this is debug code so whatever you like.

Reply:
👍 Fixed in: d3e59a29b